File: src/Backend/Services/PayrollService.cs
using KsaHrApi.Models;
using KsaHrApi.Data;
using Microsoft.EntityFrameworkCore;

namespace KsaHrApi.Services;

public interface IPayrollService
{
    Task<PayrollRecord> ProcessPayroll(Guid employeeId, int month, int year);
    Task<IEnumerable<PayrollRecord>> ProcessBulkPayroll(int month, int year);
    Task<PayrollRecord?> GetPayrollRecord(Guid employeeId, int month, int year);
}

public class PayrollService : IPayrollService
{
    private readonly HrDbContext _context;
    private readonly IGosiCalculationService _gosiService;

    public PayrollService(HrDbContext context, IGosiCalculationService gosiService)
    {
        _context = context;
        _gosiService = gosiService;
    }

    public async Task<PayrollRecord> ProcessPayroll(Guid employeeId, int month, int year)
    {
        var employee = await _context.Employees
            .FirstOrDefaultAsync(e => e.Id == employeeId && e.EmploymentStatus == "Active");

        if (employee == null)
            throw new InvalidOperationException($"Active employee not found with ID: {employeeId}");

        // Check if payroll already exists
        var existing = await _context.PayrollRecords
            .FirstOrDefaultAsync(p => p.EmployeeId == employeeId && p.PayrollMonth == month && p.PayrollYear == year);

        if (existing != null)
            return existing;

        // Calculate GOSI base (Basic Salary + Housing Allowance)
        decimal gosiBase = employee.BasicSalary + employee.HousingAllowance;

        // Calculate GOSI contributions
        var gosiResult = _gosiService.CalculateGosiContributions(employee, gosiBase);

        // Calculate gross salary
        decimal grossSalary = employee.TotalSalary;

        // Calculate total deductions
        decimal totalDeductions = gosiResult.EmployeeContribution;

        // Calculate net salary
        decimal netSalary = grossSalary - totalDeductions;

        var payrollRecord = new PayrollRecord
        {
            Id = Guid.NewGuid(),
            EmployeeId = employeeId,
            PayrollMonth = month,
            PayrollYear = year,
            BasicSalary = employee.BasicSalary,
            HousingAllowance = employee.HousingAllowance,
            TransportationAllowance = employee.TransportationAllowance,
            OtherAllowances = employee.OtherAllowances,
            OvertimePay = 0m, // To be calculated separately
            GrossSalary = grossSalary,
            GosiCalculationBase = gosiBase,
            GosiEmployeeContribution = gosiResult.EmployeeContribution,
            GosiEmployerContribution = gosiResult.EmployerContribution,
            GosiTotalContribution = gosiResult.TotalContribution,
            OtherDeductions = 0m,
            TotalDeductions = totalDeductions,
            NetSalary = netSalary,
            PaymentStatus = "Pending",
            CreatedAt = DateTime.UtcNow
        };

        _context.PayrollRecords.Add(payrollRecord);
        await _context.SaveChangesAsync();

        return payrollRecord;
    }

    public async Task<IEnumerable<PayrollRecord>> ProcessBulkPayroll(int month, int year)
    {
        var activeEmployees = await _context.Employees
            .Where(e => e.EmploymentStatus == "Active")
            .ToListAsync();

        var payrollRecords = new List<PayrollRecord>();

        foreach (var employee in activeEmployees)
        {
            try
            {
                var record = await ProcessPayroll(employee.Id, month, year);
                payrollRecords.Add(record);
            }
            catch (Exception ex)
            {
                // Log error but continue processing
                Console.WriteLine($"Error processing payroll for employee {employee.EmployeeCode}: {ex.Message}");
            }
        }

        return payrollRecords;
    }

    public async Task<PayrollRecord?> GetPayrollRecord(Guid employeeId, int month, int year)
    {
        return await _context.PayrollRecords
            .Include(p => p.Employee)
            .FirstOrDefaultAsync(p => p.EmployeeId == employeeId && p.PayrollMonth == month && p.PayrollYear == year);
    }
}

File: src/Frontend/ksa-hr-angular18/src/app/core/services/payroll.service.ts
// Payroll Service - With GOSI Integration
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { PayrollRecord } from '../models/employee.model';
import { GosiCalculationService } from './gosi-calculation.service';
import { environment } from '../../../environments/environment';

export interface PayrollProcessRequest {
  employeeId: string;
  month: number;
  year: number;
}

export interface BulkPayrollProcessRequest {
  month: number;
  year: number;
}

@Injectable({
  providedIn: 'root'
})
export class PayrollService {
  private apiUrl = `${environment.apiUrl}/Payroll`;

  constructor(
    private http: HttpClient,
    private gosiService: GosiCalculationService
  ) {}

  /**
   * Process payroll for single employee
   */
  processPayroll(request: PayrollProcessRequest): Observable<PayrollRecord> {
    return this.http.post<PayrollRecord>(`${this.apiUrl}/process`, request);
  }

  /**
   * Process bulk payroll for all active employees
   */
  processBulkPayroll(request: BulkPayrollProcessRequest): Observable<PayrollRecord[]> {
    return this.http.post<PayrollRecord[]>(`${this.apiUrl}/process-bulk`, request);
  }

  /**
   * Get payroll record
   */
  getPayrollRecord(employeeId: string, month: number, year: number): Observable<PayrollRecord> {
    const params = new HttpParams()
      .set('employeeId', employeeId)
      .set('month', month.toString())
      .set('year', year.toString());
    
    return this.http.get<PayrollRecord>(this.apiUrl, { params });
  }

  /**
   * Get payroll records by employee
   */
  getPayrollRecordsByEmployee(employeeId: string): Observable<PayrollRecord[]> {
    const params = new HttpParams().set('employeeId', employeeId);
    return this.http.get<PayrollRecord[]>(`${this.apiUrl}/employee`, { params });
  }

  /**
   * Get payroll records by month/year
   */
  getPayrollRecordsByPeriod(month: number, year: number): Observable<PayrollRecord[]> {
    const params = new HttpParams()
      .set('month', month.toString())
      .set('year', year.toString());
    
    return this.http.get<PayrollRecord[]>(`${this.apiUrl}/period`, { params });
  }

  /**
   * Calculate payroll preview (client-side)
   */
  calculatePayrollPreview(
    basicSalary: number,
    housingAllowance: number,
    transportAllowance: number,
    otherAllowances: number,
    overtimePay: number,
    isSaudiNational: boolean,
    isNewEmployee: boolean,
    gosiRegistrationDate: Date | null
  ): {
    grossSalary: number;
    gosiCalculationBase: number;
    gosiEmployeeContribution: number;
    gosiEmployerContribution: number;
    gosiTotalContribution: number;
    totalDeductions: number;
    netSalary: number;
  } {
    // Calculate gross salary
    const grossSalary = basicSalary + housingAllowance + transportAllowance + 
                        otherAllowances + overtimePay;

    // Calculate GOSI contributions
    const currentDate = new Date();
    const gosiContribution = this.gosiService.calculateGosiContributions(
      isSaudiNational,
      isNewEmployee,
      gosiRegistrationDate,
      basicSalary,
      housingAllowance,
      currentDate.getMonth() + 1,
      currentDate.getFullYear()
    );

    // Total deductions
    const totalDeductions = gosiContribution.employeeContribution;

    // Net salary
    const netSalary = grossSalary - totalDeductions;

    return {
      grossSalary: this.roundToTwoDecimals(grossSalary),
      gosiCalculationBase: gosiContribution.calculationBase,
      gosiEmployeeContribution: gosiContribution.employeeContribution,
      gosiEmployerContribution: gosiContribution.employerContribution,
      gosiTotalContribution: gosiContribution.totalContribution,
      totalDeductions: this.roundToTwoDecimals(totalDeductions),
      netSalary: this.roundToTwoDecimals(netSalary)
    };
  }

  /**
   * Generate payslip PDF
   */
  generatePayslipPdf(payrollRecordId: string): Observable<Blob> {
    return this.http.get(`${this.apiUrl}/${payrollRecordId}/payslip`, {
      responseType: 'blob'
    });
  }

  /**
   * Export payroll to Excel
   */
  exportPayrollToExcel(month: number, year: number): Observable<Blob> {
    const params = new HttpParams()
      .set('month', month.toString())
      .set('year', year.toString());
    
    return this.http.get(`${this.apiUrl}/export`, {
      params,
      responseType: 'blob'
    });
  }

  /**
   * Get payroll statistics
   */
  getPayrollStatistics(month: number, year: number): Observable<{
    totalEmployees: number;
    totalGrossSalary: number;
    totalNetSalary: number;
    totalGosiEmployee: number;
    totalGosiEmployer: number;
    averageSalary: number;
  }> {
    const params = new HttpParams()
      .set('month', month.toString())
      .set('year', year.toString());
    
    return this.http.get<any>(`${this.apiUrl}/statistics`, { params });
  }

  private roundToTwoDecimals(value: number): number {
    return Math.round(value * 100) / 100;
  }
}
